File: rust/src/days.rs
mod day01;
mod day02;
mod day03;
mod day04;
mod day05;
mod day06;
mod day07;
mod day08;
mod day09;
mod day10;
mod day11;
mod day12;
mod day13;
mod day14;
mod day15;
mod day16;
mod day17;
mod day18;
mod day19;
mod day20;
mod day21;
mod day22;
mod day23;
mod day24;
mod day25;


use std::path::PathBuf;
use crate::Result;

pub type Part = &'static (dyn Fn(&PathBuf) -> Result<String> + Send + Sync);
pub type Day = (Part, Part);
pub type Days = [Day; 25];

pub const ALL_DAYS: Days = [
        (&day01::pt1, &day01::pt2),
        (&day02::pt1, &day02::pt2),
        (&day03::pt1, &day03::pt2),
        (&day04::pt1, &day04::pt2),
        (&day05::pt1, &day05::pt2),
        (&day06::pt1, &day06::pt2),
        (&day07::pt1, &day07::pt2),
        (&day08::pt1, &day08::pt2),
        (&day09::pt1, &day09::pt2),
        (&day10::pt1, &day10::pt2),
        (&day11::pt1, &day11::pt2),
        (&day12::pt1, &day12::pt2),
        (&day13::pt1, &day13::pt2),
        (&day14::pt1, &day14::pt2),
        (&day15::pt1, &day15::pt2),
        (&day16::pt1, &day16::pt2),
        (&day17::pt1, &day17::pt2),
        (&day18::pt1, &day18::pt2),
        (&day19::pt1, &day19::pt2),
        (&day20::pt1, &day20::pt2),
        (&day21::pt1, &day21::pt2),
        (&day22::pt1, &day22::pt2),
        (&day23::pt1, &day23::pt2),
        (&day24::pt1, &day24::pt2),
        (&day25::pt1, &day25::pt2),
];

File: elixir/lib/days.ex
defmodule Days do
  @moduledoc """
  Documentation for `AOC`.
  """

  @input_dir "../inputs/"

  alias Days.{
    Day01, Day02, Day03, Day04, Day05,
    Day06, Day07, Day08, Day09, Day10,
    Day11, Day12, Day13, Day14, Day15,
    Day16, Day17, Day18, Day19, Day20,
    Day21, Day22, Day23, Day24, Day25,
  }

  # Could I generate this list with a macro?
  @soln_functions [
    {&Day01.pt1/1, &Day01.pt2/1},
    {&Day02.pt1/1, &Day02.pt2/1},
    {&Day03.pt1/1, &Day03.pt2/1},
    {&Day04.pt1/1, &Day04.pt2/1},
    {&Day05.pt1/1, &Day05.pt2/1},
    {&Day06.pt1/1, &Day06.pt2/1},
    {&Day07.pt1/1, &Day07.pt2/1},
    {&Day08.pt1/1, &Day08.pt2/1},
    {&Day09.pt1/1, &Day09.pt2/1},
    {&Day10.pt1/1, &Day10.pt2/1},
    {&Day11.pt1/1, &Day11.pt2/1},
    {&Day12.pt1/1, &Day12.pt2/1},
    {&Day13.pt1/1, &Day13.pt2/1},
    {&Day14.pt1/1, &Day14.pt2/1},
    {&Day15.pt1/1, &Day15.pt2/1},
    {&Day16.pt1/1, &Day16.pt2/1},
    {&Day17.pt1/1, &Day17.pt2/1},
    {&Day18.pt1/1, &Day18.pt2/1},
    {&Day19.pt1/1, &Day19.pt2/1},
    {&Day20.pt1/1, &Day20.pt2/1},
    {&Day21.pt1/1, &Day21.pt2/1},
    {&Day22.pt1/1, &Day22.pt2/1},
    {&Day23.pt1/1, &Day23.pt2/1},
    {&Day24.pt1/1, &Day24.pt2/1},
    {&Day25.pt1/1, &Day25.pt2/1},
  ]

  defp inputs do
    input_dir = File.ls!(@input_dir)
    1..25
      |> Enum.map(&num_to_input_name/1)
      |> Enum.filter(fn input_name -> Enum.member?(input_dir, input_name) end)
      |> Enum.map(fn name -> @input_dir <> name end)
  end

  defp num_to_input_name(i) do
    x = Integer.to_string(i)
    if String.length(x) == 1 do
      "0" <> x <> ".txt"
    else
      x <> ".txt"
    end
  end

  @doc """
  Runs all Advent of Code solutions.

  ## Examples

    iex> Days.main()

  """
  def main(_args \\ []) do
    get_day_output = fn {day_name, file_name, {pt1, pt2}} ->
      "Day #{day_name}:\n  Part 1: #{pt1.(file_name)}\n  Part 2: #{pt2.(file_name)}\n"
    end

    [1..25, inputs(), @soln_functions]
      |> Enum.zip()
      |> Enum.map(&(Task.async(fn -> get_day_output.(&1) end)))
      |> Enum.map(&Task.await/1)
      |> Enum.each(&IO.puts/1)
  end
end
